src/crimson/os/seastore/nvmedevice/nvmedevice.h

```diff
@@ -0,0 +1,329 @@
+   */
+  virtual nvme_command_ertr::future<> pass_through_io(
+    NVMePassThroughCommand& command) { return nvme_command_ertr::now(); }
+  virtual nvme_command_ertr::future<> pass_admin(
+    nvme_admin_command_t& command) { return nvme_command_ertr::now(); }
+
+  /*
+   * End-to-End Data Protection
+   *
+   * NVMe device keeps track of data integrity similar with checksum. Client can
+   * offload checksuming to NVMe device to reduce its CPU utilization
+   */
+   virtual write_ertr::future<> protected_write(
+    uint64_t offset,
+    bufferptr &bptr,
+    uint16_t stream = 0) { return write_ertr::now(); }
+
+  /*
+   * Data Health
+   *
+   * Returns list of LBAs which have almost corrupted data. Data of the LBAs
+   * will be corrupted very soon. Caller can overwrite, unmap or refresh data to
+   * protect data
+   */
+   virtual nvme_command_ertr::future<> get_data_health(
+     std::list<uint64_t>& fragile_lbas) { return nvme_command_ertr::now(); }
+
+  /*
+   * Recovery Level
+   *
+   * Regulate magnitude of SSD-internal data recovery. Caller can get good read
+   * latency with lower magnitude.
+   */
+   virtual nvme_command_ertr::future<> set_data_recovery_level(
+     uint32_t level) { return nvme_command_ertr::now(); }
+
+  /*
+   * Predictable Latency
+   *
+   * NVMe device can guarantee IO latency within pre-defined time window. This
+   * functionality will be analyzed soon.
+   */
+};
+
+/*
+ * Implementation of NVMeBlockDevice with POSIX APIs
+ *
+ * NormalNBD provides NVMe SSD interfaces through POSIX APIs which is generally
+ * available at most operating environment.
+ */
+class NormalNBD : public NVMeBlockDevice {
+public:
+  NormalNBD() {}
+  ~NormalNBD() override {}
+
+  open_ertr::future<> open(
+    const std::string &in_path,
+    seastar::open_flags mode) override;
+
+  write_ertr::future<> write(
+    uint64_t offset,
+    bufferptr &bptr,
+    uint16_t stream = 0) override;
+
+  read_ertr::future<> read(
+    uint64_t offset,
+    bufferptr &bptr) override;
+
+  nvme_command_ertr::future<> pass_through_io(
+    NVMePassThroughCommand& command) override;
+
+  nvme_command_ertr::future<> pass_admin(
+    nvme_admin_command_t& command) override;
+
+  seastar::future<> close() override;
+
+private:
+  seastar::file_desc fd = seastar::file_desc::from_fd(-1);
+  std::vector<seastar::file_desc> stream_fd;
+  nvme_version_t protocol_version;
+  bool support_multistream = false;
+  std::vector<::io_context_t> ctx;
+  std::thread completion_poller;
+  bool exit = false;
+
+  uint32_t write_life_not_set = 0;
+  uint32_t write_life_max = 1;
+
+  nvme_command_ertr::future<> identify_controller(
+    identify_controller_data_t& controller_data);
+  nvme_command_ertr::future<> identify_namespace(
+    identify_namespace_data_t& namespace_data);
+  void open_for_io(const std::string& in_path, seastar::open_flags mode);
+};
+
+
+class TestMemory : public NVMeBlockDevice {
```

Comment:
I think the definitions  for this are in memory.cc.  Can we just stick the definitions into nvmdblockdevice.cc as well?

Reply:
OK